docs/StyleGuide.md

```diff
@@ -0,0 +1,1142 @@
+# Network Style Guide
+
+This style guide is based on [consensus][consensus-style-guide].  Our guiding
+principles and enforcement through PR reviews are the same, so please read the
+relevant sections of that document, we just repeat here the most important
+topics.  However our formatting rules are slightly different.
+
+## Guiding principles
+
+We value the following principles in the networking team:
+
+* __Optimise for clarity__
+* __Consistency__: inconsistent style, especially within a single module,
+  looks sloppy, inspires little confidence in the quality of the code,
+  and distracts. Consistency is also a helpful guiding factor when deciding
+  on style guidelines in the first place; sometimes a choice between formatting
+  something this way or that way seems arbitrary in isolation, but becomes
+  clearer when seen as part of a coherent whole.
+
+## Formatting
+
+We now list the formatting rules we have converged on. As these have grown
+organically, not all code follows these rules. When touching some existing code,
+we in general recommend sticking to the existing style, but when it differs from
+the rules below, it is good practice to update the code's style to match them.
+
+2. __Indentation__: we indent by 2 spaces.
+
+   *Why:* to avoid wasting horizontal screen space.
+
+   Some consequences of this rule:
+
+   a. The `where` clause of a function body is indented 2 spaces from the left
+      margin, and the function body is indented 2 spaces from the `where`:
+
+      ```haskell
+      foo x y z =
+          ..
+        where
+          a = ..
+      ```
+
+      The `where` keyword acts as a separator between the body and the bindings.
+      Keeping them at the same indentation level would make it hard to see where
+      the body ends.
+
+      We stick with this indentation even if the `where` clause is not present,
+      just to avoid unnecessary changes when a `where` clause is added.
+
+   b. We indent record `data` and `newtype` definitions as follows:
+
+      ```haskell
+      data Foo = Foo
+          { fooBar      :: Int
+          , fooArgument :: Bool
+          }
+        deriving (Show, Eq)
+
+      newtype Foo = Foo { unFoo :: Int }
+        deriving (Show, Eq)
+      ```
+
+      The `deriving` is indented from the left margin, and the constructors
+      are indented from the `deriving` clause. This provides a consistent
+      style for data types with multiple constructors (see below).
+
+      Multiple deriving clauses using `DerivingStrategies` are aligned:
+
+      ```haskell
+      data Foo = Foo
+          { fooBar      :: Int
+          , fooArgument :: Bool
+          }
+        deriving stock    (Show, Eq, Generic)
+        deriving anyclass (NoThunks, NFData)
+
+      newtype Foo = Foo {
+            unFoo :: Int
+          }
+        deriving stock   (Show)
+        deriving newtype (Eq)
+        deriving NoThunks via InspectHeapNamed "Foo" Foo
```

Comment:
No particular ordering on these right?

Reply:
good question.  The order should be dictated by the type classes, not the `deriving` clause syntax, hence leaving it unspecified makes more sense to me.  What do you think?

Reply:
I agree!